openfe/protocols/restraint_utils/geometry/flatbottom.py

```diff
@@ -0,0 +1,137 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/openfe
+"""
+Restraint Geometry classes
+
+TODO
+----
+* Add relevant duecredit entries.
+"""
+import pathlib
+from typing import Union, Optional
+import numpy as np
+from openmm import app
+from openff.units import unit
+from openff.models.types import FloatQuantity
+import MDAnalysis as mda
+from MDAnalysis.analysis.base import AnalysisBase
+from MDAnalysis.lib.distances import calc_bonds
+
+from .harmonic import (
+    DistanceRestraintGeometry,
+)
+
+from .utils import _get_mda_topology_format, _get_mda_selection
+
+
+class FlatBottomDistanceGeometry(DistanceRestraintGeometry):
+    """
+    A geometry class for a flat bottom distance restraint between two groups
+    of atoms.
+    """
+    well_radius: FloatQuantity["nanometer"]
+
+
+class COMDistanceAnalysis(AnalysisBase):
+    """
+    Get a timeseries of COM distances between two AtomGroups
+
+    Parameters
+    ----------
+    group1 : MDAnalysis.AtomGroup
+      Atoms defining the first centroid.
+    group2 : MDANalysis.AtomGroup
+      Atoms defining the second centroid.
+    """
+    _analysis_algorithm_is_parallelizable = False
+
+    def __init__(self, group1, group2, **kwargs):
+        super().__init__(group1.universe.trajectory, **kwargs)
+
+        self.ag1 = group1
+        self.ag2 = group2
+
+    def _prepare(self):
+        self.results.distances = np.zeros(self.n_frames)
+
+    def _single_frame(self):
+        com_dist = calc_bonds(
+            self.ag1.center_of_mass(),
+            self.ag2.center_of_mass(),
+            box=self.ag1.universe.dimensions,
+        )
+        self.results.distances[self._frame_index] = com_dist
+
+    def _conclude(self):
+        pass
+
+
+def get_flatbottom_distance_restraint(
+    topology: Union[str, app.Topology],
+    trajectory: Union[str, pathlib.Path],
+    host_atoms: Optional[list[int]] = None,
+    guest_atoms: Optional[list[int]] = None,
+    host_selection: Optional[str] = None,
+    guest_selection: Optional[str] = None,
+    padding: unit.Quantity = 0.5 * unit.nanometer,
+) -> FlatBottomDistanceGeometry:
+    """
+    Get a FlatBottomDistanceGeometry by analyzing the COM distance
+    change between two sets of atoms.
+
+    The ``well_radius`` is defined as the maximum COM distance plus
+    ``padding``.
+
+    Parameters
+    ----------
+    topology : Union[str, app.Topology]
+      A topology defining the system.
+    trajectory : Union[str, pathlib.Path]
+      A coordinate trajectory for the system.
+    host_atoms : Optional[list[int]]
+      A list of host atoms indices. Either ``host_atoms`` or
+      ``host_selection`` must be defined.
+    guest_atoms : Optional[list[int]]
+      A list of guest atoms indices. Either ``guest_atoms`` or
+      ``guest_selection`` must be defined.
+    host_selection : Optional[str]
```

Comment:
Yeah I totally agree with you - to be fair even a Boresch restraint doesn't need to define host/guest. ComponentA/ComponentB might actually be better since the user supplied indices relate back to a Component.

That being said, I also agree with you that it's probably not critical for a first pass. I'm going to open an issue and we should be able to quickly refactor this out after we're done with the basics of everything else.